python_modules/dagster/dagster/_core/definitions/external_asset_graph.py

```diff
@@ -122,106 +93,214 @@ def from_repository_handles_and_external_asset_nodes(
         repo_handle_external_asset_nodes: Sequence[Tuple[RepositoryHandle, "ExternalAssetNode"]],
         external_asset_checks: Sequence["ExternalAssetCheck"],
     ) -> "ExternalAssetGraph":
-        upstream: Dict[AssetKey, AbstractSet[AssetKey]] = {}
-        source_asset_keys: Set[AssetKey] = set()
-        partitions_defs_by_key: Dict[AssetKey, Optional[PartitionsDefinition]] = {}
-        partition_mappings_by_key: Dict[AssetKey, Dict[AssetKey, PartitionMapping]] = defaultdict(
-            defaultdict
-        )
-        group_names_by_key = {}
-        freshness_policies_by_key = {}
-        auto_materialize_policies_by_key = {}
-        backfill_policies_by_key = {}
-        keys_by_atomic_execution_unit_id: Dict[str, Set[AssetKeyOrCheckKey]] = defaultdict(set)
         repo_handles_by_key = {
             node.asset_key: repo_handle
             for repo_handle, node in repo_handle_external_asset_nodes
             if not node.is_source or node.is_observable
         }
-        job_names_by_key = {
-            node.asset_key: node.job_names
-            for _, node in repo_handle_external_asset_nodes
-            if not node.is_source or node.is_observable
+
+        non_source_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        observable_source_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        non_observable_source_node_pairs: List[Tuple[RepositoryHandle, "ExternalAssetNode"]] = []
+        for repo_handle, node in repo_handle_external_asset_nodes:
+            if node.is_source and node.is_observable:
+                observable_source_node_pairs.append((repo_handle, node))
+            elif node.is_source:
+                non_observable_source_node_pairs.append((repo_handle, node))
+            else:
+                non_source_node_pairs.append((repo_handle, node))
+
+        asset_nodes_by_key = {}
+
+        # It is possible for multiple nodes to exist that share the same key. This is invalid if
+        # more than one node is materializable or if more than one node is observable. It is valid
+        # if there is at most one materializable node and at most one observable node, with all
+        # other nodes unexecutable. The asset graph will receive only a single `ExternalAssetNode`
+        # representing the asset. This will always be the materializable node if one exists; then
+        # the observable node if it exists; then finally the first-encountered unexecutable node.
+        for repo_handle, node in non_source_node_pairs:
```

Comment:
This is a super consequential decision and we need to consider it carefully. This is a pretty deep assumption that will spread throughout the system, all the way to the frontend and our top-level APIs.

The case that directly exposes this is when an asset is materializable in one repo but observable in another. What is the behavior in that case?